src/shape.rs

```diff
@@ -660,237 +677,214 @@ impl ShapeLine {
                                         fitting_start,
                                         word_range_width,
                                         number_of_blanks,
-                                    ));
+                                    );
+                                    visual_lines.push(current_visual_line);
+                                    current_visual_line = VisualLine::default();
+
                                     number_of_blanks = 0;
-                                    fit_x = line_width - glyph_size;
-                                    word_range_width = glyph_size;
+                                    fit_x = line_width - glyph_width;
+                                    word_range_width = glyph_width;
                                     fitting_start = (i, glyph_i + 1);
                                 }
                             }
                         } else {
                             // Wrap::Word
-                            let mut prev_word_width = None;
-                            if word.blank && number_of_blanks > 0 {
-                                // current word causing a wrap is a space so we ignore it
-                                number_of_blanks -= 1;
-                            } else if let Some(previous_word) = span.words.get(i - 1) {
-                                // Current word causing a wrap is not whitespace, so we ignore the
-                                // previous word if it's a whitespace
-                                if previous_word.blank {
-                                    number_of_blanks -= 1;
-                                    prev_word_width = Some(previous_word.x_advance * font_size);
+                            let mut trialing_space_width = None;
+                            if i > 0 {
+                                if let Some(previous_word) = span.words.get(i - 1) {
+                                    // Current word causing a wrap is not whitespace, so we ignore the
+                                    // previous word if it's a whitespace
+                                    if previous_word.blank {
+                                        trialing_space_width =
+                                            Some(previous_word.x_advance * font_size);
+                                        number_of_blanks = number_of_blanks.saturating_sub(1);
+                                    }
                                 }
                             }
-                            if let Some(width) = prev_word_width {
-                                word_ranges.push((
-                                    (i, 0),
+                            if let Some(width) = trialing_space_width {
+                                self.add_to_visual_line(
+                                    &mut current_visual_line,
+                                    span_index,
+                                    (i + 2, 0),
```

Comment:
Sorry for asking so many questions, but I'm just curious, since this differs so much from the congruent direction case. Why do we use `i + 2` here, but `i` in the other direction?

Reply:
My confusion was cleared up here: https://github.com/pop-os/cosmic-text/pull/108#discussion_r1134197269.